cmd/skaffold/app/cmd/build_test.go

```diff
@@ -26,47 +26,73 @@ import (
-		return []build.Artifact{{
-			ImageName: "gcr.io/skaffold/example",
-			Tag:       "test",
+	singleArtifactRunner := func(context.Context, io.Writer) ([]build.Result, error) {
+		return []build.Result{{
+			Target: latest.Artifact{
+				ImageName: "gcr.io/skaffold/example",
+			},
+			Result: build.Artifact{
+				ImageName: "gcr.io/skaffold/example",
+				Tag:       "gcr.io/skaffold/example:test",
+			},
 		}}, nil
 	}
 
-	defer func(f func(context.Context, io.Writer) ([]build.Artifact, error)) {
-		createRunnerAndBuildFunc = f
-	}(createRunnerAndBuildFunc)
+	doubleArtifactRunner := func(context.Context, io.Writer) ([]build.Result, error) {
+		return []build.Result{
+			{
+				Target: latest.Artifact{
+					ImageName: "gcr.io/skaffold/image1",
+				},
+				Result: build.Artifact{
+					ImageName: "gcr.io/skaffold/image1",
+					Tag:       "gcr.io/skaffold/image1:tag",
+				},
+			},
+			{
+				Target: latest.Artifact{
+					ImageName: "gcr.io/skaffold/image2",
+				},
+				Error: errors.New("build error"),
+			},
+		}, nil
+	}
+
+	originalBuildFormatFlag := buildFormatFlag
 
 	var tests = []struct {
 		description    string
 		template       string
 		expectedOutput []byte
-		mock           func(context.Context, io.Writer) ([]build.Artifact, error)
+		mock           func(context.Context, io.Writer) ([]build.Result, error)
 		shouldErr      bool
 	}{
 		{
-			description:    "quiet flag print build images with no template",
-			expectedOutput: []byte(`{"builds":[{"imageName":"gcr.io/skaffold/example","tag":"test"}]}`),
-			shouldErr:      false,
-			mock:           mockCreateRunner,
+			description:    "single image with no template",
+			expectedOutput: []byte(`{"builds":[{"imageName":"gcr.io/skaffold/example","tag":"gcr.io/skaffold/example:test"}]}`),
+			mock:           singleArtifactRunner,
 		},
 		{
-			description:    "quiet flag print build images applies pattern specified in template ",
+			description:    "single image with specified template",
 			template:       "{{range .Builds}}{{.ImageName}} -> {{.Tag}}\n{{end}}",
-			expectedOutput: []byte("gcr.io/skaffold/example -> test\n"),
-			shouldErr:      false,
-			mock:           mockCreateRunner,
+			expectedOutput: []byte("gcr.io/skaffold/example -> gcr.io/skaffold/example:test\n"),
+			mock:           singleArtifactRunner,
+		},
+		{
+			description: "build errors out when incorrect template specified",
+			template:    "{{.Incorrect}}",
+			shouldErr:   true,
+			mock:        singleArtifactRunner,
 		},
 		{
-			description:    "build errors out when incorrect template specified",
-			template:       "{{.Incorrect}}",
-			expectedOutput: nil,
-			shouldErr:      true,
-			mock:           mockCreateRunner,
+			description:    "two images, no template, one error",
+			expectedOutput: []byte(`{"builds":[{"imageName":"gcr.io/skaffold/image1","tag":"gcr.io/skaffold/image1:tag"}]}`),
+			mock:           doubleArtifactRunner,
```

Comment:
this should err right (see my above comment)

Reply:
I'm not sure I agree (see my above comment about templating ;))

Reply:
i feel the overall build command should result in an error. 
We can decide on how to write the command output later. I can make the necessary changes in the deploy flow if we decide to show errors.

Reply:
@balopat ^^ do you agree?

Reply:
yes, fair enough, let's fail the build command when there is a failure, that is orthogonal to writing the results or not.

Reply:
updated